src/types.rs

```diff
@@ -0,0 +1,627 @@
+
+                impl From<&$typ> for AstarteDataTypeIndividual {
+                    fn from(d: &$typ) -> Self {
+                        AstarteDataTypeIndividual {
+                            individual_data: Some(IndividualData::$astartedatatype(d.clone().into())),
+                            }
+                    }
+                }
+        )*
+    };
+}
+
+impl_type_conversion_traits!({
+    (f64, AstarteDouble),
+    (i32, AstarteInteger),
+    (bool, AstarteBoolean),
+    (i64, AstarteLongInteger),
+    (&str, AstarteString),
+    (String, AstarteString),
+    (Vec<u8>, AstarteBinaryBlob),
+});
+
+impl From<DateTime<Utc>> for AstarteDataTypeIndividual {
+    fn from(value: DateTime<Utc>) -> Self {
+        let system_time: SystemTime = value.into();
+
+        AstarteDataTypeIndividual {
+            individual_data: Some(IndividualData::AstarteDateTime(system_time.into())),
+        }
+    }
+}
+
+impl From<Vec<f64>> for AstarteDataTypeIndividual {
+    fn from(value: Vec<f64>) -> Self {
+        AstarteDataTypeIndividual {
+            individual_data: Some(IndividualData::AstarteDoubleArray(AstarteDoubleArray {
+                astarte_double: value,
+            })),
+        }
+    }
+}
+
+impl From<Vec<i32>> for AstarteDataTypeIndividual {
+    fn from(value: Vec<i32>) -> Self {
+        AstarteDataTypeIndividual {
+            individual_data: Some(IndividualData::AstarteIntegerArray(AstarteIntegerArray {
+                astarte_integer: value,
+            })),
+        }
+    }
+}
+
+impl From<Vec<i64>> for AstarteDataTypeIndividual {
+    fn from(value: Vec<i64>) -> Self {
+        AstarteDataTypeIndividual {
+            individual_data: Some(IndividualData::AstarteLongIntegerArray(
+                AstarteLongIntegerArray {
+                    astarte_long_integer: value,
+                },
+            )),
+        }
+    }
+}
+
+impl From<Vec<bool>> for AstarteDataTypeIndividual {
+    fn from(value: Vec<bool>) -> Self {
+        AstarteDataTypeIndividual {
+            individual_data: Some(IndividualData::AstarteBooleanArray(AstarteBooleanArray {
+                astarte_boolean: value,
+            })),
+        }
+    }
+}
+
+impl From<Vec<String>> for AstarteDataTypeIndividual {
+    fn from(value: Vec<String>) -> Self {
+        AstarteDataTypeIndividual {
+            individual_data: Some(IndividualData::AstarteStringArray(AstarteStringArray {
+                astarte_string: value,
+            })),
+        }
+    }
+}
+
+impl From<Vec<Vec<u8>>> for AstarteDataTypeIndividual {
+    fn from(value: Vec<Vec<u8>>) -> Self {
+        AstarteDataTypeIndividual {
+            individual_data: Some(IndividualData::AstarteBinaryBlobArray(
+                AstarteBinaryBlobArray {
+                    astarte_binary_blob: value,
+                },
+            )),
+        }
+    }
+}
+
+impl From<Vec<DateTime<Utc>>> for AstarteDataTypeIndividual {
+    fn from(value: Vec<DateTime<Utc>>) -> Self {
```

Comment:
The line above says this is not covered by test. Given it's the only non-trivial conversion, I'd add a test for it